File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/BpaPdcStream/PhasorDefinition.vb
'*******************************************************************************************************
'  PhasorDefinition.vb - PDCstream Phasor definition
'  Copyright © 2008 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2008
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Text

Namespace BpaPdcStream

    <CLSCompliant(False), Serializable()> _
    Public Class PhasorDefinition

        Inherits PhasorDefinitionBase

        Private m_ratio As Single
        Private m_calFactor As Single
        Private m_shunt As Single
        Private m_voltageReferenceIndex As Int32

        Protected Sub New()
        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

            ' Deserialize phasor definition
            m_ratio = info.GetSingle("ratio")
            m_calFactor = info.GetSingle("calFactor")
            m_shunt = info.GetSingle("shunt")
            m_voltageReferenceIndex = info.GetInt32("voltageReferenceIndex")

        End Sub

        Public Sub New(ByVal parent As ConfigurationCell)

            MyBase.New(parent)

        End Sub

        Public Sub New(ByVal parent As ConfigurationCell, ByVal index As Int32, ByVal entryValue As String)

            MyBase.New(parent)

            Dim entry As String() = entryValue.Split(","c)
            Dim entryType As String = UCase(Left(Trim(entry(0)), 1))
            Dim defaultPhasor As PhasorDefinition

            If Not parent Is Nothing Then
                Dim configFile As ConfigurationFrame = Me.Parent.Parent

                If entryType = "V" Then
                    [Type] = PhasorType.Voltage
                    defaultPhasor = configFile.DefaultPhasorV
                ElseIf entryType = "I" Then
                    [Type] = PhasorType.Current
                    defaultPhasor = configFile.DefaultPhasorI
                Else
                    [Type] = PhasorType.Voltage
                    defaultPhasor = configFile.DefaultPhasorV
                End If
            Else
                defaultPhasor = New PhasorDefinition(DirectCast(Nothing, ConfigurationCell))
            End If

            If entry.Length > 1 Then Ratio = CDbl(Trim(entry(1))) Else Ratio = defaultPhasor.Ratio
            If entry.Length > 2 Then CalFactor = CDbl(Trim(entry(2))) Else ConversionFactor = defaultPhasor.ConversionFactor
            If entry.Length > 3 Then Offset = CDbl(Trim(entry(3))) Else Offset = defaultPhasor.Offset
            If entry.Length > 4 Then Shunt = CDbl(Trim(entry(4))) Else Shunt = defaultPhasor.Shunt
            If entry.Length > 5 Then VoltageReferenceIndex = CInt(Trim(entry(5))) Else VoltageReferenceIndex = defaultPhasor.VoltageReferenceIndex
            If entry.Length > 6 Then Label = Trim(entry(6)) Else Label = defaultPhasor.Label

            Me.Index = index

        End Sub

        Public Sub New(ByVal parent As ConfigurationCell, ByVal phasorDefinition As IPhasorDefinition)

            MyBase.New(parent, phasorDefinition)

        End Sub

        Public Overrides ReadOnly Property DerivedType() As System.Type
            Get
                Return Me.GetType
            End Get
        End Property

        Public Shadows ReadOnly Property Parent() As ConfigurationCell
            Get
                Return MyBase.Parent
            End Get
        End Property

        ' BPA PDCstream uses a custom conversion factor (see shared overload below)
        Public Overrides Property ConversionFactor() As Single
            Get
                Return 1.0F
            End Get
            Set(ByVal value As Single)
                ' Ignore updates...
            End Set
        End Property

        Public Property Ratio() As Single
            Get
                Return m_ratio
            End Get
            Set(ByVal value As Single)
                m_ratio = value
            End Set
        End Property

        Public Property CalFactor() As Single
            Get
                Return m_calFactor
            End Get
            Set(ByVal value As Single)
                m_calFactor = value
            End Set
        End Property

        Public Property Shunt() As Single
            Get
                Return m_shunt
            End Get
            Set(ByVal value As Single)
                m_shunt = value
            End Set
        End Property

        Public Property VoltageReferenceIndex() As Int32
            Get
                Return m_voltageReferenceIndex
            End Get
            Set(ByVal value As Int32)
                m_voltageReferenceIndex = value
            End Set
        End Property

        Public Overloads Shared ReadOnly Property ConversionFactor(ByVal phasor As PhasorDefinition) As Single
            Get
                If phasor.Type = PhasorType.Voltage Then
                    Return phasor.CalFactor * phasor.Ratio
                Else
                    Return phasor.CalFactor * phasor.Ratio / phasor.Shunt
                End If
            End Get
        End Property

        Public Shared ReadOnly Property ConfigFileFormat(ByVal phasor As PhasorDefinition) As String
            Get
                With New StringBuilder
                    Select Case phasor.Type
                        Case PhasorType.Voltage
                            .Append("V"c)
                        Case PhasorType.Current
                            .Append("I"c)
                    End Select

                    .Append("," & _
                        phasor.Ratio & "," & _
                        phasor.CalFactor & "," & _
                        phasor.Offset & "," & _
                        phasor.Shunt & "," & _
                        phasor.VoltageReferenceIndex & "," & _
                        phasor.Label)

                    Return .ToString()
                End With
            End Get
        End Property

        Public Overrides ReadOnly Property MaximumLabelLength() As Int32
            Get
                Return Int32.MaxValue
            End Get
        End Property

        Protected Overrides ReadOnly Property BodyLength() As UInt16
            Get
                Return 0
            End Get
        End Property

        Protected Overrides ReadOnly Property BodyImage() As Byte()
            Get
                Throw New NotImplementedException("PDCstream does not include phasor definition in descriptor packet - must be defined in external INI file")
            End Get
        End Property

        Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

            MyBase.GetObjectData(info, context)

            ' Serialize phasor definition
            info.AddValue("ratio", m_ratio)
            info.AddValue("calFactor", m_calFactor)
            info.AddValue("shunt", m_shunt)
            info.AddValue("voltageReferenceIndex", m_voltageReferenceIndex)

        End Sub

        Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
            Get
                Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

                baseAttributes.Add("Ratio", m_ratio)
                baseAttributes.Add("CalFactor", m_calFactor)
                baseAttributes.Add("Shunt", m_shunt)
                baseAttributes.Add("Voltage Reference Index", m_voltageReferenceIndex)

                Return baseAttributes
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/BpaPdcStream/PhasorDefinition.cs
//*******************************************************************************************************
//  PhasorDefinition.vb - PDCstream Phasor definition
//  Copyright © 2008 - TVA, all rights reserved - Gbtc
//
//  Build Environment: VB.NET, Visual Studio 2008
//  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
//      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
//       Phone: 423/751-2827
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - J. Ritchie Carroll
//       Initial version of source generated
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.Serialization;
using System.Security.Permissions;

namespace PhasorProtocols.BpaPdcStream
{
    /// <summary>
    /// Represents the BPA PDCstream implementation of a <see cref="IPhasorDefinition"/>.
    /// </summary>
    [Serializable()]
    public class PhasorDefinition : PhasorDefinitionBase
    {
        #region [ Members ]

        // Fields
        private double m_ratio;
        private double m_calFactor;
        private double m_shunt;
        private int m_voltageReferenceIndex;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="PhasorDefinition"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="PhasorDefinition"/>.</param>
        public PhasorDefinition(IConfigurationCell parent)
            : base(parent)
        {
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinition"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="ConfigurationCell"/> parent of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="label">The label of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="scale">The integer scaling value of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="offset">The offset of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="type">The <see cref="PhasorType"/> of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="voltageReference">The associated <see cref="IPhasorDefinition"/> that represents the voltage reference (if any).</param>
        public PhasorDefinition(ConfigurationCell parent, string label, uint scale, double offset, PhasorType type, PhasorDefinition voltageReference)
            : base(parent, label, scale, offset, type, voltageReference)
        {
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinition"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="ConfigurationCell"/> parent of this <see cref="PhasorDefinition"/>.</param>
        /// <param name="index">Index of phasor within INI based configuration file.</param>
        /// <param name="entryValue">The entry value from the INI based configuration file.</param>
        public PhasorDefinition(ConfigurationCell parent, int index, string entryValue)
            : base(parent)
        {
            string[] entry = entryValue.Split(',');
            string entryType = entry[0].Trim().Substring(0, 1).ToUpper();
            PhasorDefinition defaultPhasor;

            if (parent != null)
            {
                ConfigurationFrame configFile = this.Parent.Parent;

                if (entryType == "V")
                {
                    PhasorType = PhasorType.Voltage;
                    defaultPhasor = configFile.DefaultPhasorV;
                }
                else if (entryType == "I")
                {
                    PhasorType = PhasorType.Current;
                    defaultPhasor = configFile.DefaultPhasorI;
                }
                else
                {
                    PhasorType = PhasorType.Voltage;
                    defaultPhasor = configFile.DefaultPhasorV;
                }
            }
            else
            {
                defaultPhasor = new PhasorDefinition(null as ConfigurationCell);
            }

            if (entry.Length > 1)
                Ratio = double.Parse(entry[1].Trim());
            else
                Ratio = defaultPhasor.Ratio;

            if (entry.Length > 2)
                CalFactor = double.Parse(entry[2].Trim());
            else
                ConversionFactor = defaultPhasor.ConversionFactor;

            if (entry.Length > 3)
                Offset = double.Parse(entry[3].Trim());
            else
                Offset = defaultPhasor.Offset;

            if (entry.Length > 4)
                Shunt = double.Parse(entry[4].Trim());
            else
                Shunt = defaultPhasor.Shunt;

            if (entry.Length > 5)
                VoltageReferenceIndex = (int)double.Parse(entry[5].Trim());
            else
                VoltageReferenceIndex = defaultPhasor.VoltageReferenceIndex;

            if (entry.Length > 6)
                Label = entry[6].Trim();
            else
                Label = defaultPhasor.Label;

            this.Index = index;
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinition"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected PhasorDefinition(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize phasor definition
            m_ratio = info.GetSingle("ratio");
            m_calFactor = info.GetSingle("calFactor");
            m_shunt = info.GetSingle("shunt");
            m_voltageReferenceIndex = info.GetInt32("voltageReferenceIndex");
        }


        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets the <see cref="ConfigurationCell"/> parent of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public virtual new ConfigurationCell Parent
        {
            get
            {
                return base.Parent as ConfigurationCell;
            }
            set
            {
                base.Parent = value;
            }
        }

        /// <summary>
        /// Gets or sets the conversion factor of this <see cref="ChannelDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// This method is hidden from the consumer since BPA PDCstream uses a custom conversion factor.
        /// Property always returns 1.0, updates are ignored.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public override double ConversionFactor
        {
            // BPA PDCstream uses a custom conversion factor (see shared overload below)
            get
            {
                return 1.0D;
            }
            set
            {
                // Ignore updates...
            }
        }

        /// <summary>
        /// Gets or sets ratio of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public double Ratio
        {
            get
            {
                return m_ratio;
            }
            set
            {
                m_ratio = value;
            }
        }

        /// <summary>
        /// Gets or sets calibration factor of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public double CalFactor
        {
            get
            {
                return m_calFactor;
            }
            set
            {
                m_calFactor = value;
            }
        }

        /// <summary>
        /// Gets or sets shunt value of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public double Shunt
        {
            get
            {
                return m_shunt;
            }
            set
            {
                m_shunt = value;
            }
        }

        /// <summary>
        /// Gets or sets voltage reference index of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public int VoltageReferenceIndex
        {
            get
            {
                return m_voltageReferenceIndex;
            }
            set
            {
                m_voltageReferenceIndex = value;
            }
        }

        /// <summary>
        /// Gets the maximum length of the <see cref="ChannelDefinitionBase.Label"/> of this <see cref="PhasorDefinition"/>.
        /// </summary>
        public override int MaximumLabelLength
        {
            get
            {
                return 256;
            }
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="PhasorDefinition"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Ratio", m_ratio.ToString());
                baseAttributes.Add("CalFactor", m_calFactor.ToString());
                baseAttributes.Add("Shunt", m_shunt.ToString());
                baseAttributes.Add("Voltage Reference Index", m_voltageReferenceIndex.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize phasor definition
            info.AddValue("ratio", m_ratio);
            info.AddValue("calFactor", m_calFactor);
            info.AddValue("shunt", m_shunt);
            info.AddValue("voltageReferenceIndex", m_voltageReferenceIndex);
        }

        #endregion

        #region [ Static ]

        // Static Methods

        // Calculates a BPA PDCstream phasor's custom conversion factor
        internal static double CustomConversionFactor(PhasorDefinition phasor)
        {
            if (phasor.PhasorType == PhasorType.Voltage)
                return phasor.CalFactor * phasor.Ratio;

            return phasor.CalFactor * phasor.Ratio / phasor.Shunt;
        }

        // Creates phasor information for an INI based BPA PDCstream configuration file
        internal static string ConfigFileFormat(IPhasorDefinition definition)
        {
            PhasorDefinition phasor = definition as PhasorDefinition;

            if (phasor != null)
                return (phasor.PhasorType == PhasorType.Voltage ? "V" : "I") + "," + phasor.Ratio + "," + phasor.CalFactor + "," + phasor.Offset + "," + phasor.Shunt + "," + phasor.VoltageReferenceIndex + "," + phasor.Label;

            return "";
        }

        // Delegate handler to create a new BPA PDCstream phasor definition
        internal static IPhasorDefinition CreateNewDefinition(IConfigurationCell parent, byte[] binaryImage, int startIndex, out int parsedLength)
        {
            IPhasorDefinition phasorDefinition = new PhasorDefinition(parent);

            parsedLength = phasorDefinition.Initialize(binaryImage, startIndex, 0);

            return phasorDefinition;
        }

        #endregion        
    }
}